### DISCUSSION TITLE
Level crossing extensions in NewGRF

### DISCUSSION THREAD

**Discussion by PikkaBird** [OpenTTD/OpenTTD#12939] [Features]
At the moment, the options for drawing different level crossings are extremely limited. Rail and Roadtypes share the same 5 variables, which are:

40 Terrain type.
41 Enhanced tunnels.
42 Level crossing status.
43 Construction date.
44 Town zone.

![Wreadingstone Transport, 24th Nov 1901](https://github.com/user-attachments/assets/5b6d30dc-c92d-4b73-867f-a0002a90d126)

Of these variables, only 40 and 44 are actually useful. 41 is unimplemented, 43 just returns the current date for non-depot tiles, and 42 doesn't differentiate between an open crossing and a non-crossing. This means it is not possible to draw different crossings based either on year of construction, or on the roadtype (a dirt farm track crossing should surely look different from a crossing on a major highway?).

I would like to discuss the following changes/additions to the spec, to allow more flexibility in drawing custom level crossings:

1) a new variable, which reports the roadtype to the rail tile, and the railtype to the road tile. This would double as an unambiguous crossing detector (since it would only return a non-FF value on crossing tiles).*

2) altering var 42 to return a different value for open vs no crossing.

3) extending var 43 to return the actual construction date for level crossings (and potentially every road/rail tile).

Of these three options, 2 seems like it could cause problems with existing GRFs, and in any case would be redundant if 1 were implemented. 3 would be nice-to-have, but would involve storing more data for each road/rail tile. But **1, adding the cross-type-reporting, seems like a no-brainer and would get us 95% of the way to all our level crossing desires**. Anyone up for implementing it? Or can see any issues with the idea?

--

_*Presumably _two_ new variables for each type, because we also have tramtypes (so road vars read rail + tram, tram reads rail + road, rail reads road + tram), but ignoring that for now, for the sake of keeping the discussion simple._

**Comment by PikkaBird** [comment]
![Pluntown Transport, 26th Feb 1952](https://github.com/user-attachments/assets/116a8b6b-7357-4e53-a5fe-cb8253a2907a)

![Sesleigh Transport, 15th Nov 1960](https://github.com/user-attachments/assets/d8323f1d-4388-4d18-8504-60c1efdd3fac)

Next issue - the fact that RVs drive a considerable distance onto the crossing tile before stopping is annoying, from a graphic design point, eg, it makes drawing things like gates or barriers problematic. What are the potential issues of changing this behaviour to have vehicles stop at the edge of the tile?

**Comment by PikkaBird** [comment]
As pointed out by PeterN on Discord, the behavior of stopping forward on the tile allows a standard-length RV to fit between two parallel tracks, reducing the chances of RVs being hit by a train while waiting for a crossing. However, since #9931, RVs no longer stop between crossing tiles anyway, so it's no longer an issue. #9931 also highlights that altering the behavior of level crossings has _not_ met significant backlash from the player base.

**Comment by PeterN** [comment]
As for 3, there is no space to store a 32 bit construction date in the map, and level crossings don't have off-map data.

**Comment by frosch123** [comment]
Option 1 is implemented in #13934.

**Reply by yozora3** [reply]
I love you

**Comment by LC-Zorg** [comment]
> What are the potential issues of changing this behaviour to have vehicles stop at the edge of the tile?

Before version 13.0 it made sense that a standard length vehicle could hide between two tracks to avoid collisions. After adding the multi-track level crossing function this is no longer the case, because vehicles will no longer stop in the middle of the crossing.

The issue may be extended crossing time. Players who know how to avoid collisions at crossings know how far away they can place a signal. But when there is a lot of traffic, it is placed as close to the crossing as possible. This has the consequence that trains can pass right behind the vehicles, and this is intentional. If you move the starting point away, on refined networks it can lead to many accidents and the need to change the position of the signal, which may not be easy.
I think that for most players such a change would not make a big difference, but for those more involved it could be frustrating.
It would probably be best if each set with roads could specify a different stopping point for vehicles.

**Reply by PikkaBird** [reply]
> The issue may be extended crossing time.

Both multi-track level crossings and articulated road vehicles extend crossing time, typically to a much greater extent than shifting a stopping point 1/4 of a tile. As far as I'm aware it has never been raised as an issue in the three and eighteen years respectively that these features have existed.
